### PR TITLE
Fix sha512crypt-opencl failing self tests on NVIDIA OpenCL driver 361

### PR DESCRIPTION
**Pull request by claudioandre**
Is this a reasonable solution for #2215 ?

It works on super (which uses CUDA 8). I bet no new problems introduced.
****

In this change:
1. if there is a known problem, we do something.
2. otherwise, let the compiler do its best.

### CODE REVIEW

**Comment by magnumripper** [comment]
IMHO it's a bit messy as of now (as in hard to follow). For instance, NVIDIA_STUPID_BUG_1 is **always** defined for any nvidia. Then if it's defined **and** we are within a version span, we work around bugs. Doesn't really make sense.

I think something along the lines of this would be easier to understand/maintain:

opencl_misc.h
```c
#if gpu_nvidia(DEVICE_INFO) && DEV_VER_MAJOR >= 361
/* some comment on this particular bug */
#define NVIDIA_STUPID_BUG_1    1
#endif

#if gpu_nvidia(DEVICE_INFO) && DEV_VER_MAJOR >= 310 && DEV_VER_MAJOR <= 313
/* some comment on this particular bug */
#define NVIDIA_STUPID_BUG_2    1
#endif

(...)

#if gpu_amd(DEVICE_INFO) && DEV_VER_MAJOR >= this && DEV_VER_MAJOR <= that
/* some comment on this particular bug */
#define AMD_STUPID_BUG_1    1
#endif

(...)
```

some_kernels.cl
```c
(...)
#if AMD_STUPID_BUG_1
      #pragma unroll 2 /* or whatever workaround */
#elif NVIDIA_STUPID_BUG_1
      #pragma unroll 16 /* or whatever workaround */
#endif
(...)
```

**Comment by magnumripper** [comment]
BTW determining a correct or non-ambiguous "CUDA version" from OpenCL output can be tricky.

The stuff @solardiz installed on super was called "CUDA 7.5". According to --list=cuda-devices, it's CUDA runtime 7.5 and CUDA driver 8.0 and according to --list=opencl-devices, it's "CUDA 8.0.13".

**Comment by claudioandre** [comment]
> IMHO it's a bit messy as of now (as in hard to follow). For instance, NVIDIA_STUPID_BUG_1 is always defined for any nvidia. Then if it's defined and we are within a version span, we work around bugs. Doesn't really make sense.

I can see your point, but there are some things I though when I first did this:
* avoid messing with shared code and opencl_misc.h is used a lot.
* keep driver versions inside the .cl file, easy to find and change - probably only to me.
* allow an easy way to solve a problem if one happens (just change the .cl file in the right place). Consider the situation when an user opened 2215.
* I called them bugs, but in fact I was only documenting (somehow) my findings. Seems, nobody else faced problems

But it is fair, I can start to document my findings in a better way.

**Comment by claudioandre** [comment]
> according to --list=opencl-devices, it's "CUDA 8.0.13"

Strange (since I never did myself anything NVIDIA related I will say nothing). 
* But there is a printf inside NVIDIA software outputing this.

**Comment by magnumripper** [comment]
> avoid messing with shared code and opencl_misc.h is used a lot.

Fair enough, if you believe a bug only affects your code you can keep it all within your kernel. Still, it could be something like:
```c
#if gpu_nvidia(DEVICE_INFO) && DEV_VER_MAJOR >= 361
/* some comment on this particular bug, what device, issue # or whatever */
#define NVIDIA_UNROLL_BUG    1
#endif

#if gpu_amd(DEVICE_INFO) && DEV_VER_MAJOR >= this && DEV_VER_MAJOR <= that
/* some comment on this particular bug, what device, issue # or whatever */
#define AMD_UNROLL_BUG    1
#endif

(...)
#if AMD_UNROLL_BUG
      #pragma unroll 2
#elif NVIDIA_UNROLL_BUG
      #pragma unroll 16
#endif
```

This applies to myself too; my documentation of bugs has varied between poor and good.

**Comment by magnumripper** [comment]
Some more data points:
- Titan X w/ 361.42 also fails without the explicit unroll.
- Titan X w/ 352.63 works fine with or without explicit unroll.
- R9 290X w/ 1800.11 works fine with or without explicit unroll.

**Comment by claudioandre** [comment]
> Titan X w/ 361.42 also fails without the explicit unroll.
Titan X w/ 352.63 works fine with or without explicit unroll.

As it used to work on super in the past.

> R9 290X w/ 1800.11 works fine with or without explicit unroll.

Only saw fails on VLIW5 (remember, I tested on well boards too).

> so it's better to over-use it than not.

Do you mean enable the workaround always?

**Comment by magnumripper** [comment]
>> so it's better to over-use it than not.

> Do you mean enable the workaround always?

Perhaps not, but we should err to the safe side so we rather over-use it than miss some, eg:
```c
#if gpu_nvidia(DEVICE_INFO) && DEV_VER_MAJOR > 352
      // So far any version above 352 we've tested needed this to work
      #pragma unroll 16
#endif
```

**Comment by magnumripper** [comment]
Looks fine to me now. We may need to improve it over time of course, as vendors develop more bugs...
